flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -409,6 +404,22 @@ public MetricGroup addGroup(String name) {
 		return addGroup(name, false);
 	}
 
+	@Override
+	public MetricGroup addGroup(String key, String value) {
```

Comment:
is there a case that the previous approach didn't cover?

Reply:
This change wants to make the fallback behavior more reasonable. 

When user uses this method, he will want the key be a key and the value be a value.

But in the previous approach, if the parent group is key group, that will leads to key becomes to value and value becomes the new key. 

I think it is more acceptable to make these unexpected behavior fall back to ordinary metric group. 

What do you think?

Reply:
Wouldn't it be cleaner to override `addGroup(String key, String value)` in `GenericKeyMetricGroup` instead?

Reply:
Yes, you are right. I will refine it. Thank you.

Reply:
I remove the first if-else statement by overriding the `GenericKeyMetricGroup#addGroup(String key, String value)`, but remain the second one.
Because when the first group is already created as `GenericMetricGroup` then the second group will be created as `GenericKeyMetricGroup`.
It seems that there is no way to be cleaner, unless we change the `keyed` be a enum type and move the decision to the private `addGroup` method.

Reply:
For any `GenericKeyMetricGroup` to be created, `keyed` must be true. Since this is only the case when creating the first group I don't understand how it could happen that the second group is created as a key group.

This commit https://github.com/zentol/flink/commit/f3a91f6296994a33a7be448bec40f261b5343972 removes the special case in addGroup(String, String) and reverts the logic for `createChildGroup` (which introduced a bug). The `MetricGroupTest` is passing.

Reply:
But in the blow case:

Reply:
In the below case, the `subGroup2` will get a `GenericValueMetricGroup` back.
```
val subGroup = group.addGroup("key", "value")
val subGroup2 = group.addGroup("key").addGroup("value")
```
I prefer not to change the behavior of original API, so I think it is better to return a `GenericMetricGroup`.

Reply:
Sorry for the last incorrect comment.
This change will make `GenerciMetricGroup` always create `GenericKeyMetricGroup`, because I also change the behavior of `#AbstractMetricGroup#createChildGroup(String name)`.
It seems there is a bug in `MetricGroupTest#alwaysCanCreateGenericMetricGroup`, because what I want is always return exactly `GenericMetricGroup` and child class is unacceptable.

Reply:
The reasont why I reverted the previous implementation is that I want to make `MetricGroup#addGroup(String name)` always create `GenericMetricGroup`, even through the class type of `group` is `GenericKeyMericGroup`.

Reply:
ok, your solution makes sense. 👍